File: Elixir/Language Basics/patternmatching.ex
defmodule Patternmatching do
@moduledoc """
Erlang and Elixir for Imperative Programmers, 2016
Chapter 14 - Pattern Matching
"""

  def always_return_42 do
    42
  end

  def always_return_42_short, do: 42

  def return_42_cond_when_true(flag) do
    cond do
      flag ->
        42
      1 ->
        0
    end
  end

  def return_42_case_when_true(flag) do
    case flag do
      true -> 42
      _ -> 0
    end
  end

  def incorrect_case(a,b) do
      case a == b do
        true -> 0
        false -> -1
      end
  end

  @spec incorrect_case_spec(atom(),atom()) :: number()
  def incorrect_case_spec(a,b) do
    case a == b do
      true -> 0
      false -> -1
    end
  end

  def incorrect_case_guards(a,b) when is_atom(a) and is_atom(b) do
    case a == b do
      true -> 0
      false -> -1
    end
  end

  def case_with_guards(a,b) when is_number(a) and is_number(b) do
    case a == b do
      true -> true
      _ -> false
    end
  end

  def case_with_guards(a,b) when is_atom(a) and is_atom(b) do
    case a == b do
      true -> true
      _ -> false
    end
  end

  def call() do
    incorrect_case(1,1)
  end

end

File: Erlang/Language Basics/patternmatching.erl
%%%-------------------------------------------------------------------
%%% @author Wolfgang Loder
%%% @doc
%%% Erlang and Elixir for Imperative Programmers, 2016
%%% Chapter 14 - Pattern Matching
%%% @end
%%%-------------------------------------------------------------------
-module(patternmatching).
-author("Wolfgang Loder").

-export([always_return_42/0, return_42_if_when_true/1,return_42_case_when_true/1,incorrect_case/2,guard_function/1]).

always_return_42() ->
	42.

return_42_if_when_true(Flag) ->
  if
    Flag =:= true ->
      42;
    true ->
      0
  end.

return_42_case_when_true(Flag) ->
  case Flag of
    true -> 42;
    _ -> 0
  end.

incorrect_case(a,b) ->
  case a == b of
    true -> -1;
    false -> 0
  end;
incorrect_case(a,Y) ->
  case Y > 0 of
    true -> 42;
    false -> 0
  end;
incorrect_case(_,Y) ->
  case Y > 0 of
    true -> 43;
    false -> 0
  end.

guard_function(Flag) when Flag == true -> 42;
guard_function(Flag) when Flag == false -> 0.
